File: ios/Runner/Api/AlbumHostApiImpl.swift
//
//  AlbumHostApiImpl.swift
//  Runner
//
//  Created by t k on 2024/08/09.
//

class AlbumHostApiImpl: AlbumHostApi {
    private let repository = AlbumRepository()
    
    func getAllAlbums() throws -> [AlbumDTO] {
        let albums = repository.findAll()
        
        return albums.map { $0.toPigeon() }
    }

    func getAlbumsByArtistId(artistId: String) throws -> [AlbumDTO] {
        let albums = repository.findByArtistId(artistId: ArtistId(id: artistId))
        
        return albums.map { $0.toPigeon() }
    }

    func getRecentAlbums(count: Int64) throws -> [AlbumDTO] {
        let albums = repository.findRecentAlbums(count: Int(count))
        
        return albums.map { $0.toPigeon() }
    }

    func getAlbumById(id: String) throws -> AlbumDTO? {
        guard let album = repository.findByAlbumId(albumId: AlbumId(id: id)) else {
            return nil
        }

        return album.toPigeon()
    }
}

File: android/app/src/main/kotlin/dev/tcode/thinmpf/api/AlbumHostApiImpl.kt
package dev.tcode.thinmpf.api

import AlbumDTO
import AlbumHostApi
import android.content.Context
import dev.tcode.thinmpf.model.valueObject.AlbumId
import dev.tcode.thinmpf.model.valueObject.ArtistId
import dev.tcode.thinmpf.repository.AlbumRepository
import dev.tcode.thinmpf.extension.toPigeon

class AlbumHostApiImpl(context: Context) : AlbumHostApi {
    private val repository = AlbumRepository(context)

    override fun getAllAlbums(): List<AlbumDTO> {
        val albums = repository.findAll()

        return albums.map { it.toPigeon() }
    }

    override fun getAlbumsByArtistId(artistId: String): List<AlbumDTO> {
        val albums = repository.findByArtistId(ArtistId(artistId))

        return albums.map { it.toPigeon() }
    }

    override fun getRecentAlbums(count: Long): List<AlbumDTO> {
        val albums = repository.findRecentAlbums(count.toInt())

        return albums.map { it.toPigeon() }
    }

    override fun getAlbumById(id: String): AlbumDTO? {
        val album = repository.findByAlbumId(AlbumId(id))

        return album?.toPigeon()
    }
}